### PR TITLE
chore: add structurally typed GenericPythonConfiguration

### PR DESCRIPTION
**Pull request by hoodmane**
Use structural typing for the return value of get_python_configurations so that we don't have to specify a union of all possible `PythonConfiguration` types.

This makes it slightly more convenient to add a new platform.

### CODE REVIEW

**Review by henryiii — APPROVED** [review]

File: cibuildwheel/typing.py [hunk 1]
```diff
@@ -42,3 +42,9 @@
 
 PlatformName = Literal["linux", "macos", "windows"]
 PLATFORMS: Final[set[PlatformName]] = {"linux", "macos", "windows"}
+
+
+class GenericPythonConfiguration(Protocol):
+    @property
+    def identifier(self) -> str:
+        pass
```
Comment (henryiii) [inline cibuildwheel/typing.py:50]:
```suggestion
        ...
```

Typing uses `...` conventionally.

File: cibuildwheel/__main__.py [hunk 2]
```diff
@@ -341,11 +346,7 @@ def print_preamble(platform: str, options: Options, identifiers: list[str]) -> N
 def get_build_identifiers(
     platform: PlatformName, build_selector: BuildSelector, architectures: set[Architecture]
 ) -> list[str]:
-    python_configurations: (
-        list[cibuildwheel.linux.PythonConfiguration]
-        | list[cibuildwheel.windows.PythonConfiguration]
-        | list[cibuildwheel.macos.PythonConfiguration]
-    )
+    python_configurations: typing.Sequence[GenericPythonConfiguration]
```
Comment (henryiii) [inline cibuildwheel/__main__.py:349]:
```suggestion
    python_configurations: Sequence[GenericPythonConfiguration]
```

Comment (henryiii) [reply cibuildwheel/__main__.py:349]: Ahh, can't add `from collections.abc import Sequence` as a suggestion. This is fine for now, but ideally we should do that (also `Set`, from there, for input arguments, is better than a concrete and mutable `set`).

Comment (hoodmane) [reply cibuildwheel/__main__.py:349]: oh yeah it hadn't occurred to me that `set[A]` is similar to `MutableSet[A]`? but still more specific? But `Dict` is the same as `dict` and `Map` / `MutableMap` are different?

Comment (henryiii) [reply cibuildwheel/__main__.py]: `Dict` is just a older way to say `dict[...]`. I think `typing.Set` / `set` are also similar, while `collections.abc.{Map,MutableMap,Set,MutableSet}` are more general structural types. Input parameters should generally be as loose as possible, and output parameters should be as specific as possible (usually concrete).

**Pull request merged**